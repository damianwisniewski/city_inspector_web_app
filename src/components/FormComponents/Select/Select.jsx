import React, { Component } from 'react'
import PropTypes from 'prop-types'

import './Select.scss'

class Select extends Component {
	static option = ({ children, ...rest }) => <option {...rest}>{children}</option>

	state = {
		focused: false,
		selected: '',
	}

	componentDidMount() {
		document.addEventListener('mousedown', this.toggleSelectFocus)
	}

	componentWillUnmount() {
		document.removeEventListener('mousedown', this.toggleSelectFocus)
	}

	toggleSelectFocus = e => {
		e.preventDefault()
		if (this.field.contains(e.target) || this.label.contains(e.target)) {
			this.setState({ focused: true })
		} else {
			this.setState({ focused: false })
		}
	}

	setValue = e => {
		const { onSelectOption } = this.props
		this.setState({ selected: e.target.textContent })
		if (onSelectOption) onSelectOption(e)
	}

	render() {
		const { id, children, label, ...rest } = this.props
		const { focused, selected } = this.state
		const focusClass = focused ? 'input-wrapper__field--focused' : ''

		return (
			<div className='input-wrapper input-wrapper--select'>
				<div className={`input-wrapper__field input-wrapper__field--select ${focusClass}`}>
					<select
						{...rest}
						className='field__output'
						ref={field => (this.field = field)}
						id={id}
						value={selected}
						onChange={() => {}}
					>
						{children}
					</select>
					<ul data-amount={3} className='input-wrapper__select-list'>
						{children.map((option, index) => (
							<li data-type='list-item' onClick={this.setValue} key={option + '_' + index}>
								{option.props.children}
							</li>
						))}
					</ul>
				</div>
				<label ref={label => (this.label = label)} className='input-wrapper__label' htmlFor={id}>
					{label}
				</label>
			</div>
		)
	}
}

Select.defaultProps = {
	label: '',
}

Select.propTypes = {
	id: PropTypes.string.isRequired,
	label: PropTypes.string,
	onSelectOption: PropTypes.func.isRequired,
}

export default Select
