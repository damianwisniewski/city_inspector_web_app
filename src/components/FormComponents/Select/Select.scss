@import '../../../assets/styles/index';

.input-wrapper--select {
	z-index: 50;
	position: relative;

	&::after {
		content: '\f107';
		font-family: 'Font Awesome 5 Free';
		font-size: 18px;
		font-weight: 900;
		position: absolute;
		z-index: 1;
		right: 20px;
		top: 50%;
		transform: translateY(-50%);
	}

	.field__output {
		position: relative;
		z-index: 1;
		font-size: 16px;
		background-color: transparent;
		margin: 0;
		width: 100%;
		height: 100%;
		display: flex;
		justify-content: flex-start;
		align-items: center;
		padding: 0 10px;
		border: none;
		appearance: none;

		&::-ms-expand {
			display: none;
		}

		&:invalid {
			background-color: $white;
			border: 1px solid transparent;
		}
	}

	.input-wrapper__select-list {
		position: absolute;
		top: 33px;
		left: 50%;
		z-index: -1;
		border-bottom-left-radius: 3px;
		border-bottom-right-radius: 3px;
		background-color: $white;
		color: $dark_gray;
		box-shadow: $shadow;
		width: 98%;
		list-style: none;
		margin: 0;
		padding: 0;
		opacity: 0;
		overflow: hidden;
		visibility: hidden;
		border-top: none;
		max-height: 200px;
		transform: translateX(-50%) translateY(-35px);
		transition: opacity 0.3s linear, transform 0.3s linear, visibility 0.3s linear;
		transition-delay: 0.1s;

		& li {
			padding: 5px 10px;
		}
	}

	div.input-wrapper__field--select {
		padding: 0;
	}

	div.input-wrapper__field--focused {
		border: 1px solid $primary_blue;
		box-shadow: $shine;

		& li:hover {
			background-color: $primary_blue;
			color: $white;
		}

		& ~ .input-wrapper__label {
			color: $blue;
		}

		& .input-wrapper__select-list {
			visibility: visible;
			opacity: 1;
			transform: translateX(-50%) translateY(0%);
		}
	}
}
