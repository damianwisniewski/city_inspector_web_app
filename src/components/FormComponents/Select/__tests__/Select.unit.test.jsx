import React from 'react'
import { configure, shallow, ShallowWrapper } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import Select from '../Select'

configure({ adapter: new Adapter() })

describe('<Select />', () => {
	/**
	 * @type {ShallowWrapper}
	 */
	let SelectElement
	const id = 'mockId'
	const onSelectOption = () => {}

	beforeEach(() => {
		SelectElement = shallow(
			<Select onSelectOption={onSelectOption} id={id}>
				<option value=''>placeholder</option>
				<option value='Value1'>Value1</option>
				<option value='Value2'>Value2</option>
				<option value='Value3'>Value3</option>
			</Select>,
			{ lifecycleExperimental: true },
		)
	})

	test('should render correctly Select', () => {
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should render correctly Select with focus state', () => {
		SelectElement.setState({ focused: true })
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should render correctly Select with selected value in state', () => {
		SelectElement.setState({ selected: 'Value1' })
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should set target.context to state selected, after click list item (option)', () => {
		SelectElement.find('[data-type="list-item"]')
			.at(1)
			.simulate('click', {
				target: {
					textContent: 'Value1',
				},
			})

		expect(SelectElement.state().selected).toBe('Value1')
	})

	test('should call onSelectOption function from props, after click list item (option)', () => {
		const spy = jest.fn()
		SelectElement.setProps({ onSelectOption: spy })
		SelectElement.find('[data-type="list-item"]')
			.at(1)
			.simulate('click', {
				target: {
					textContent: 'Value1',
				},
			})

		expect(spy).toBeCalledTimes(1)
		expect(spy).toBeCalledWith({
			target: {
				textContent: 'Value1',
			},
		})
	})
})
