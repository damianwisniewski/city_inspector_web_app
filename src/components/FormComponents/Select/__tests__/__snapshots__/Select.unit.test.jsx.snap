// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render correctly Select 1`] = `
<div
  className="input-wrapper input-wrapper--select"
>
  <div
    className="input-wrapper__field input-wrapper__field--select "
  >
    <select
      className="field__output"
      id="mockId"
      onChange={[Function]}
      onSelectOption={[Function]}
      value=""
    >
      <option
        value=""
      >
        placeholder
      </option>
      <option
        value="Value1"
      >
        Value1
      </option>
      <option
        value="Value2"
      >
        Value2
      </option>
      <option
        value="Value3"
      >
        Value3
      </option>
    </select>
    <ul
      className="input-wrapper__select-list"
      data-amount={3}
    >
      <li
        data-type="list-item"
        key="[object Object]_0"
        onClick={[Function]}
      >
        placeholder
      </li>
      <li
        data-type="list-item"
        key="[object Object]_1"
        onClick={[Function]}
      >
        Value1
      </li>
      <li
        data-type="list-item"
        key="[object Object]_2"
        onClick={[Function]}
      >
        Value2
      </li>
      <li
        data-type="list-item"
        key="[object Object]_3"
        onClick={[Function]}
      >
        Value3
      </li>
    </ul>
  </div>
  <label
    className="input-wrapper__label"
    htmlFor="mockId"
  />
</div>
`;

exports[`<Select /> should render correctly Select with focus state 1`] = `
<div
  className="input-wrapper input-wrapper--select"
>
  <div
    className="input-wrapper__field input-wrapper__field--select input-wrapper__field--focused"
  >
    <select
      className="field__output"
      id="mockId"
      onChange={[Function]}
      onSelectOption={[Function]}
      value=""
    >
      <option
        value=""
      >
        placeholder
      </option>
      <option
        value="Value1"
      >
        Value1
      </option>
      <option
        value="Value2"
      >
        Value2
      </option>
      <option
        value="Value3"
      >
        Value3
      </option>
    </select>
    <ul
      className="input-wrapper__select-list"
      data-amount={3}
    >
      <li
        data-type="list-item"
        key="[object Object]_0"
        onClick={[Function]}
      >
        placeholder
      </li>
      <li
        data-type="list-item"
        key="[object Object]_1"
        onClick={[Function]}
      >
        Value1
      </li>
      <li
        data-type="list-item"
        key="[object Object]_2"
        onClick={[Function]}
      >
        Value2
      </li>
      <li
        data-type="list-item"
        key="[object Object]_3"
        onClick={[Function]}
      >
        Value3
      </li>
    </ul>
  </div>
  <label
    className="input-wrapper__label"
    htmlFor="mockId"
  />
</div>
`;

exports[`<Select /> should render correctly Select with selected value in state 1`] = `
<div
  className="input-wrapper input-wrapper--select"
>
  <div
    className="input-wrapper__field input-wrapper__field--select "
  >
    <select
      className="field__output"
      id="mockId"
      onChange={[Function]}
      onSelectOption={[Function]}
      value="Value1"
    >
      <option
        value=""
      >
        placeholder
      </option>
      <option
        value="Value1"
      >
        Value1
      </option>
      <option
        value="Value2"
      >
        Value2
      </option>
      <option
        value="Value3"
      >
        Value3
      </option>
    </select>
    <ul
      className="input-wrapper__select-list"
      data-amount={3}
    >
      <li
        data-type="list-item"
        key="[object Object]_0"
        onClick={[Function]}
      >
        placeholder
      </li>
      <li
        data-type="list-item"
        key="[object Object]_1"
        onClick={[Function]}
      >
        Value1
      </li>
      <li
        data-type="list-item"
        key="[object Object]_2"
        onClick={[Function]}
      >
        Value2
      </li>
      <li
        data-type="list-item"
        key="[object Object]_3"
        onClick={[Function]}
      >
        Value3
      </li>
    </ul>
  </div>
  <label
    className="input-wrapper__label"
    htmlFor="mockId"
  />
</div>
`;
